src/LibraryManager/CacheService/CacheService.cs

```diff
@@ -124,28 +126,20 @@ private async Task<string> GetResourceAsync(string url, string localFile, int ex
         }
 
         /// <summary>
-        /// Refreshes the cache for the given set of files if expired 
+        /// Refreshes the cache for the given set of files if expired
         /// </summary>
-        /// <param name="librariesCacheMetadata"></param>
-        /// <param name="cancellationToken"></param>
-        /// <returns></returns>
-        public async Task RefreshCacheAsync(IEnumerable<CacheFileMetadata> librariesCacheMetadata, CancellationToken cancellationToken)
+        public async Task RefreshCacheAsync(IEnumerable<CacheFileMetadata> librariesCacheMetadata, ILogger logger, CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            List<Task> refreshTasks = new List<Task>();
+            await ParallelUtility.ForEachAsync(DownloadFileIfNecessaryAsync, MaxConcurrentDownloads, librariesCacheMetadata, cancellationToken);
 
-            foreach (CacheFileMetadata metadata in librariesCacheMetadata)
+            async Task DownloadFileIfNecessaryAsync(CacheFileMetadata metadata)
             {
-                if (!File.Exists(metadata.DestinationPath) || File.GetLastWriteTime(metadata.DestinationPath) < DateTime.Now.AddDays(-_libraryExpiresAfterDays))
+                if (!File.Exists(metadata.DestinationPath) || File.GetLastWriteTime(metadata.DestinationPath) < DateTime.Now.AddDays(-LibraryExpiresAfterDays))
                 {
-                    Task readFileTask = DownloadToFileAsync(metadata.Source, metadata.DestinationPath, attempts: 5, cancellationToken: cancellationToken);
-                    await readFileTask;
-                    refreshTasks.Add(readFileTask);
+                    logger.Log(string.Format(Resources.Text.DownloadingFile, metadata.Source), LogLevel.Operation);
```

Comment:
Just to make sure, the ILogger.Log implementation is thread safe, right?

Reply:
Yes, they all should be.